File: users/src/interactors/users_interactor.rs
use crate::entities::user::User;
use crate::ports::users_repository::UsersRepository;
use shared::ports::logger::Logger;
use uuid::Uuid;

#[derive(Clone)]
pub struct UsersInteractor {
    repository: Box<dyn UsersRepository + Send + Sync>,
    logger: Box<dyn Logger + Send + Sync>,
}

impl UsersInteractor {
    pub fn new(
        repository: Box<dyn UsersRepository + Send + Sync>,
        logger: Box<dyn Logger + Send + Sync>,
    ) -> Self {
        UsersInteractor { repository, logger }
    }

    pub fn create(&self, user: User) -> Result<(), String> {
        self.logger.debug(
            "UsersInteractor.create".to_owned(),
            format!("Creating user: {:?}", user),
        );
        self.repository.create(user)
    }

    pub fn read(&self, id: Uuid) -> Result<Option<User>, String> {
        self.logger.debug(
            "UsersInteractor.read".to_owned(),
            format!("Reading user: {}", id),
        );
        self.repository.read(id)
    }

    pub fn read_by_email(&self, email: String) -> Result<Option<User>, String> {
        self.logger.debug(
            "UsersInteractor.read_by_email".to_owned(),
            format!("Reading user by email: {}", email),
        );
        self.repository.read_by_email(email)
    }

    pub fn update(&self, user: User) -> Result<(), String> {
        self.logger.debug(
            "UsersInteractor.update".to_owned(),
            format!("Updating user: {:?}", user),
        );
        self.repository.update(user)
    }

    pub fn delete(&self, id: Uuid) -> Result<(), String> {
        self.logger.debug(
            "UsersInteractor.delete".to_owned(),
            format!("Deleting user: {}", id),
        );
        self.repository.delete(id)
    }
}

File: users/src/interactors/users.interactor.ts
import { validateIdentity } from 'src/entities/identities';
import {
  NotValidatedUserThumbnail,
  UserThumbnail,
  validateUserThumbnail,
} from 'src/entities/users.thumbnails';
import { NotValidatedUser, User, validateUser } from 'src/entities/users';
import { UsersRepository } from 'src/ports/users.repository';
import { v4 as uuidv4 } from 'uuid';

export interface UserInput {
  identities: Array<{
    system: string;
    id: string;
  }>;
  username: string;
  email: string;
}

export class UsersInteractor {
  private repository: UsersRepository;
  private logger: Logger;

  constructor({
    repository,
    logger,
  }: {
    repository: UsersRepository;
    logger: Logger;
  }) {
    this.repository = repository;
    this.logger = logger;
    this.logger.setPrefix('UsersInteractor');
  }

  async create(userInput: UserInput): Promise<User> {
    try {
      const user: NotValidatedUser = {
        ...userInput,
        createdAt: new Date(),
        identities: [
          ...(userInput.identities ?? []),
          {
            system: process.env.VTT_SYSTEM || 'vtt',
            id: uuidv4(),
          },
        ],
      };
      validateUser(user);
      const existingUser = await this.repository.readByEmail(user.email);
      if (existingUser) {
        throw new EmailAlreadyInUseError();
      }
      const savedUser = await this.repository.create(user);
      this.logger.info('create', savedUser);
      return savedUser;
    } catch (error) {
      this.logger.error('create', userInput, typeof error, error.message);
      throw error;
    }
  }

  async readUserThumbnailByIdentity(identity: {
    system: string;
    id: string;
  }): Promise<UserThumbnail | undefined> {
    try {
      validateIdentity(identity);
      const user = await this.repository.readByIdentity(identity);
      if (!user) {
        return undefined;
      }
      const userThumbnail: NotValidatedUserThumbnail = {
        username: user.username,
      };
      validateUserThumbnail(userThumbnail);
      return userThumbnail;
    } catch (error) {
      this.logger.error(
        'readThumbnailByIdentity',
        identity,
        error.message || typeof error,
      );
      throw error;
    }
  }

  async update(user: User, userInput: UserInput): Promise<User> {
    try {
      if (user.email !== userInput.email) {
        throw new UnauthorizedError();
      }
      const updatedUser = {
        ...user,
        ...userInput,
        identities: user.identities,
      };
      validateUser(updatedUser);
      this.logger.info('update', user.email, updatedUser);
      return await this.repository.update(updatedUser);
    } catch (error) {
      this.logger.error(
        'update',
        user.email,
        userInput,
        error.message || typeof error,
      );
      throw error;
    }
  }

  async delete(user: User): Promise<User> {
    try {
      this.logger.info('delete', user.email);
      return await this.repository.delete(user);
    } catch (error) {
      this.logger.error('delete', user.email, error.message || typeof error);
      throw error;
    }
  }
}

export class EmailAlreadyInUseError extends Error {}

export class UnauthorizedError extends Error {}
